Plugins/Player/Player.cpp

```diff
@@ -1336,4 +1344,163 @@ ArgumentStack Player::SetResManOverride(ArgumentStack&& args)
+                        {
+                            if (pEntry->szPlot_Id.CStr() == tag)
+                            {
+                                pCreature->m_pJournal->m_lstEntries[i] = newJournal;
+                                break;
+                            }
+                        }
+                    }
+                    if(i == entries.num)
+                        pCreature->m_pJournal->m_lstEntries.Add(newJournal);
+                    
+                    pMessage->SendServerToPlayerJournalAddQuest(pPlayer,
+                                                                    newJournal.szPlot_Id,
+                                                                    newJournal.nState,
+                                                                    newJournal.nPriority,
+                                                                    newJournal.nPictureIndex,
+                                                                    newJournal.bQuestCompleted,
+                                                                    newJournal.nCalendarDay,
+                                                                    newJournal.nTimeOfDay,
+                                                                    newJournal.szName,
+                                                                    newJournal.szText);
+                    retval =pCreature->m_pJournal->m_lstEntries.num; // Success
+                    
+                    //If no update message is desired, we can keep it silent.
+                    //Useful for restoring entries on playerlogin
+                    if(!silentUpdate)
+                    {
+                        pMessage->SendServerToPlayerJournalUpdated(pPlayer,1,newJournal.bQuestCompleted,newJournal.szName);
+                    }
+                }
+                else
+                {
+                    LOG_ERROR("Unable to get CNWSMessage");
+                }
+                
+        }
+    }
+
+    return Services::Events::Arguments(retval);
+}
+
+ArgumentStack Player::GetJournalEntry(ArgumentStack&& args)
+{
+    SJournalEntry lastJournalEntry;
+    int found = -1;
+    if (auto *pPlayer = player(args))
+    {
+        auto *pCreature = Globals::AppManager()->m_pServerExoApp->GetCreatureByGameObjectID(pPlayer->m_oidNWSObject);
+        if (pCreature && pCreature->m_pJournal)
+        {
+            auto entries = pCreature->m_pJournal->m_lstEntries;
+            const auto tag = Services::Events::ExtractArgument<std::string>(args);
+            ASSERT_OR_THROW(tag.size() > 0);
+            if (entries.num > 0)
+            {
+                auto pEntry = entries.element;
+                for (int i = 0; i < entries.num; i++, pEntry++)
+                {
+                    if (pEntry->szPlot_Id.CStr() == tag)
+                        {
+                            lastJournalEntry = entries[i];
+                            found = i;
```

Comment:
Should we abort if we find the tag, or do we need to look at the last entry with the same tag?
If there's only one with the tag, maybe move the big return here and then you don't need the `found` variable at all.

Reply:
Default journal entries add multiple entries to the player with the same tag - the last one is (as far as I can tell) the one that is generally presented/visible to the player. Even when you remove a journal entry via native RemoveJournalEntry NWScript method - it only ever removes the last entry - leaving some empty journal entries in the journal list on the player. (Somewhat memory-leakish behaviour in the base game)

The AddCustomJournal entry method attempts to avoid that sort of behaviour by overwriting entries with the same tag. This Get method should be compatible with both native journal entries and the custom entries.

Also - the found variable is used so I can return some sort of indication to the NWScript context that no valid journal entry was found - eg: no need to do more popping of return variables if the journal entry was not found etc.
Seemed more efficient to pop one return variable with -1 to indicate to the nwscript that there is no further variables to be expected etc.
If there is a better way of handling that - happy to include it.

Reply:
In that case, it would be better to start at the last element and iterate backwards until you find the first tag.

And what I mean is if you add the full data return here, you don't need the `found` variable. As soon as you hit the end of the function, you know it's not found, so you can just return `-1` directly.

Reply:
Ok - will try to refactor to start at entries.num -1 and work down towards 0 etc
Will also add the return directly to the loop